src/cascadia/TerminalApp/TerminalPage.cpp

```diff
@@ -2522,91 +2527,127 @@ namespace winrt::TerminalApp::implementation
     // - eventArgs: the PasteFromClipboard event sent from the TermControl
     fire_and_forget TerminalPage::_PasteFromClipboardHandler(const IInspectable /*sender*/,
                                                              const PasteFromClipboardEventArgs eventArgs)
+    try
     {
-        const auto data = Clipboard::GetContent();
-
-        // This will switch the execution of the function to a background (not
-        // UI) thread. This is IMPORTANT, because the getting the clipboard data
-        // will crash on the UI thread, because the main thread is a STA.
-        co_await winrt::resume_background();
+        // Counterintuitively pasting the clipboard should block the UI thread, because that's
+        // actually the expectation of users: If you type something, paste and then type something again,
+        // text should be written in exactly that order, no matter how loaded the CPU is. See GH#15315.
+        assert(Dispatcher().HasThreadAccess());
```

Comment:
I would suggest suppressing whitespace changes. The indentation changed because I moved the try/catch up to surround the entire function.